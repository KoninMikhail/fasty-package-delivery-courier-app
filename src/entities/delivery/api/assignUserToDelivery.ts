import { createEffect } from 'effector';
import { apiClient, Delivery, User } from '@/shared/api';

export type AssignUserToDeliveryParameters = {
    userId: User['id'];
    deliveryId: Delivery['id'];
};

export type AssignUserToDeliveryResponse = Delivery;

export const assignUserToDeliveryFx = createEffect<
    AssignUserToDeliveryParameters,
    AssignUserToDeliveryResponse,
    Error
>(async ({ userId, deliveryId }) => {
    return apiClient.patchDelivery(
        { courier_id: userId },
        {
            params: {
                deliveryId,
            },
        },
    );
});
