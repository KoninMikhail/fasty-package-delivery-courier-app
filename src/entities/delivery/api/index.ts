export * from './getMyDeliveriesFx';
export * from './setDeliveryStatus';
