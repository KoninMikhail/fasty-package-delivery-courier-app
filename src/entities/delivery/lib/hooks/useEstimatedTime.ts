import { calculateTimeDifferenceInMinutes } from "../utils/calculateTimeDifferenceInMinutes";
import {useEffect, useState} from "react";

export const useEstimatedTime = (date: Date) => {
    const [estimatedMinutes, setEstimatedMinutes] = useState(() =>
        calculateTimeDifferenceInMinutes(date),
    );
    useEffect(() => {
        const interval = setInterval(() => {
            setEstimatedMinutes((previous) => (previous <= 0 ? 0 : previous - 1));
        }, 10000);
        return () => clearInterval(interval);
    }, []);

    return estimatedMinutes;
};