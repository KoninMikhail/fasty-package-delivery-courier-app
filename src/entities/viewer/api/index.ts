export * from './authByEmailFx';
export * from './logoutFx';
export * from './resetByEmailRequestFx';
