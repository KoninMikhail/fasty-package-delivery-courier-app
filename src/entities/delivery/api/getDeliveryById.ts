import { createEffect } from 'effector';
import { apiClient, Delivery } from '@/shared/api';

type GetDeliveryByIdParameters = {
    deliveryId: Delivery['id'];
};

type GetDeliveryByIdResponse = Delivery;

export const getDeliveryById = createEffect<
    GetDeliveryByIdParameters,
    GetDeliveryByIdResponse,
    Error
>(async ({ deliveryId }) => {
    try {
        const respose = await apiClient.fetchDeliveryById({
            params: { deliveryId },
        });
        return respose;
    } catch {
        throw new Error('Error while fetching delivery details');
    }
});
