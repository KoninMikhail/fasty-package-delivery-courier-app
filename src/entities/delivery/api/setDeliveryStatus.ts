import { createEffect } from 'effector/effector.umd';
import { apiClient, Delivery } from '@/shared/api';

export type SetDeliveryStatusParameters = {
    id: Delivery['id'];
    comment: string;
    state: Delivery['states'];
};

export type SetDeliveryStatusResponse = Delivery;

export const setDeliveryStatus = createEffect<
    SetDeliveryStatusParameters,
    SetDeliveryStatusResponse,
    Error
>(async ({ id, state, comment }) => {
    try {
        const data = { states: state, comment };
        const parameters = {
            params: {
                deliveryId: id,
            },
        };
        return await apiClient.patchDelivery(data, parameters);
    } catch {
        throw new Error('Error while setting delivery status');
    }
});
