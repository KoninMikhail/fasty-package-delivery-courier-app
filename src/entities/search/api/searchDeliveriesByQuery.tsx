import { createEffect } from 'effector';
import { apiClient } from '@/shared/api';

export const searchDeliveriesByQueryFx = createEffect({
    handler: async (query: string) => {
        try {
            return await apiClient.searchDeliveriesByQuery({
                queries: {
                    query,
                },
            });
        } catch (error) {
            throw new Error(error);
        }
    },
});
