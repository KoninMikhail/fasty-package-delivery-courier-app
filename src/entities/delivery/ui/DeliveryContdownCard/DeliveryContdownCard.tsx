import {
    Button,
    Card,
    CardBody,
    CardFooter,
    CardHeader,
    Chip,
    Divider,
    Link,
} from '@nextui-org/react';
import { Delivery } from '@/shared/api';
import { IoCall } from 'react-icons/io5';
import { sharedServicesSubway } from '@/shared/services';
import { useTranslation } from 'react-i18next';
import { forwardRef, ReactNode } from 'react';
import { Link as ReactRouterLink } from 'react-router-dom';
import { sharedConfigRoutes } from '@/shared/config';
import { sharedLibHelpers } from '@/shared/lib';
import { getDeliveryContact } from '../../lib/utils/getDeliveryContact';
import { useEstimatedTime } from '../../lib/hooks/useEstimatedTime';
import {
    ADDRESS,
    CLIENT_NOT_FOUND,
    EXPIRED,
    TIME_LEFT,
    translationNS,
} from '../../config';
import { getDeliverySystemId } from '../../lib/utils/getDeliverySystemId';
import { getDeliveryAddress } from '../../lib/utils/getDeliveryAdress';
import { getDeliveryMetro } from '../../lib/utils/getDeliveryMetro';

const { SubwayStationWithIcon } = sharedServicesSubway;
const { RouteName } = sharedConfigRoutes;
const { getMaskedPhone } = sharedLibHelpers;
const { DELIVERIES } = RouteName;

/**
 * @name HeaderLayout
 * @description Represents the header layout of a card, showing the subway station and the countdown timer.
 */

interface HeaderLayoutProperties {
    station: ReactNode;
    countdown: ReactNode;
}

export const HeaderLayout: FunctionComponent<HeaderLayoutProperties> = ({
    station,
    countdown,
}) => (
    <div className="flex w-full items-center gap-2">
        <div className="flex-grow">{station}</div>
        <div className="flex items-center gap-1.5">{countdown}</div>
    </div>
);

/**
 * Renders the address section of the delivery card.
 */
export const AddressDisplay: FunctionComponent<{
    address: string;
}> = ({ address }) => {
    const { t } = useTranslation(translationNS);
    return (
        <div className="flex flex-col">
            <span className="text-md truncate">{`${t(ADDRESS)}:`}</span>
            <span className="text-md truncate font-bold">{address}</span>
        </div>
    );
};

/**
 * Displays the contact information of the client, including name and phone number, with a direct call action button.
 */
export const ContactDetails: FunctionComponent<{
    name?: string;
    phone?: string;
}> = ({ name, phone }) => {
    const { t } = useTranslation(translationNS);

    if (!name || !phone)
        return <div className="text-md font-bold">{t(CLIENT_NOT_FOUND)}</div>;

    const maskedPhone = getMaskedPhone(phone);

    return (
        <>
            <div className="flex-grow">
                <div className="truncate text-sm font-bold">{name}</div>
                <div className="text-sm">{maskedPhone}</div>
            </div>
            <Button
                role="link"
                isIconOnly
                color="success"
                className="text-lg text-content1"
            >
                <IoCall />
            </Button>
        </>
    );
};

/**
 * Renders a time chip that shows either the estimated time until delivery or indicates that the delivery is expired.
 */
export const DeliveryTimer: FunctionComponent<{
    date: Date;
}> = ({ date }) => {
    const { t } = useTranslation(translationNS);

    const EXPIRED_MINS_COUNT = 0;
    const MINUTES_IN_HOUR = 60;
    const CLOSE_TO_EXPIRED_THRESHOLD = 30; // Minutes before considered close to expired

    const estimatedMinutes = useEstimatedTime(date);
    const hoursLeft = Math.floor(estimatedMinutes / MINUTES_IN_HOUR);
    const minutesLeft = estimatedMinutes % MINUTES_IN_HOUR;

    const isExpired = estimatedMinutes <= EXPIRED_MINS_COUNT;
    const isCloseToExpired = estimatedMinutes <= CLOSE_TO_EXPIRED_THRESHOLD;

    if (isExpired)
        return (
            <Chip color="danger" size="sm">
                {t(EXPIRED)}
            </Chip>
        );

    return (
        <Chip color={isCloseToExpired ? 'warning' : 'default'}>
            {t(TIME_LEFT, {
                hours: hoursLeft,
                minutes: minutesLeft,
            })}
        </Chip>
    );
};

/**
 * A card component that displays a countdown to a delivery date, along with relevant delivery information.
 *
 * @param {Delivery} [delivery] - Optional. The delivery details including date, time end, address, and contact information.
 * @returns {JSX.Element} The delivery countdown card with delivery information.
 */
interface DeliveryCountdownCardProperties {
    delivery: Delivery;
}

export const DeliveryCountdownCard = forwardRef<
    HTMLDivElement,
    DeliveryCountdownCardProperties
>(({ delivery }, reference) => {
    const deadline = delivery.time_end;
    const id = getDeliverySystemId(delivery);
    const address = getDeliveryAddress(delivery);
    const contact = getDeliveryContact(delivery);
    const metro = getDeliveryMetro(delivery);

    const getDetailsPageLink = (): string => `${DELIVERIES}/${id}`;

    return (
        <Card ref={reference} className="min-w-[300px] max-w-[600px]">
            <ReactRouterLink to={getDetailsPageLink()} className="relative">
                <CardHeader className="flex gap-3">
                    <HeaderLayout
                        countdown={<DeliveryTimer date={deadline} />}
                        station={<SubwayStationWithIcon value={metro} />}
                    />
                </CardHeader>
                <Divider />

                <CardBody>
                    <AddressDisplay address={address} />
                </CardBody>
            </ReactRouterLink>
            <Divider />
            <CardFooter>
                <Link
                    href={`tel:+${contact?.phone}`}
                    className="flex w-full justify-between gap-2 text-lg text-foreground"
                >
                    <ContactDetails
                        name={contact?.name}
                        phone={contact?.phone}
                    />
                </Link>
            </CardFooter>
        </Card>
    );
});
