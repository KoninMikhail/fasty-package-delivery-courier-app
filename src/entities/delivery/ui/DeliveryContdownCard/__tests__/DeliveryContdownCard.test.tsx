import { render, screen } from "@testing-library/react";
import { beforeEach, describe, expect, it, vi } from "vitest";

import { DeliveryCountdownCard } from "../DeliveryContdownCard";
import { Delivery } from "@/shared/api";
import { useEstimatedTime } from "../../../lib/hooks/useEstimatedTime";


const mockDelivery = {
  "id": 59,
  "car": true,
  "client_id": 32,
  "comment": "beatae et consequatur autem quis reprehenderit commodi illum index.ts",
  "contact_id": 69,
  "contents": "Черта лысого получишь! хотел было.",
  "deleted": false,
  "date": "2024-02-22",
  "express": true,
  "manager_id": 4,
  "order_id": 22,
  "states": "delivering",
  "time_end": "08:06",
  "time_start": "12:10",
  "order": {
    "id": 22,
    "client_id": 95,
    "contact_id": 127,
    "address_id": 122,
    "description": "Sed rem et sit laboriosam omnis. Quas autem ipsum in ut explicabo facilis. Sunt quod alias et tempore sunt. Quia eligendi voluptatum est tenetur ipsa dolor.",
    "cost": "74200.00",
    "state": "draft",
    "payment": "pre",
    "payment_type": "cash",
    "deleted": false,
    "created_at": "2024-02-07T10:15:05.000000Z",
    "updated_at": "2024-02-25T07:27:52.000000Z",
    "submanager": null,
    "deadline": "2024-02-13T09:16:11.000000Z"
  },
  "weight": "40",
  "courier": {
    "id": 19,
    "first_name": "Виктор",
    "last_name": "Козлов",
    "gender": "male",
    "email": "kozlov.v@test.ru",
    "email_verified_at": "2024-03-04T13:12:02.000000Z",
    "deleted": false,
    "created_at": "2024-03-04T13:12:02.000000Z",
    "updated_at": "2024-03-04T13:23:32.000000Z",
    "user_role": {
      "name": "Курьер",
      "capabilities": [],
      "states": []
    }
  },
  "courier_id": 19,
  "contact": {
    "id": 69,
    "client_id": 53,
    "name": "Белозёрова Лидия Сергеевна",
    "email": "beloziorova.l@yandex.ru",
    "job": "",
    "default": false,
    "deleted": false,
    "phone": "+7(999)125-12-02"
  },
  "client": {
    "id": 32,
    "client_type": "organization",
    "name": "Голизна, ПАО",
    "deleted": false,
    "created_at": "2024-03-04T13:12:03.000000Z",
    "updated_at": "2024-03-04T13:12:03.000000Z",
    "contacts": [
      {
        "id": 39,
        "client_id": 32,
        "name": "Гаврилова Доминика Львовна",
        "email": "gavrilova.d@golizna.ru",
        "job": "Владелец",
        "default": false,
        "deleted": false,
        "phone": "+7(930)700-47-68"
      },
      {
        "id": 40,
        "client_id": 32,
        "name": "Кудрявцева Анна Романовна",
        "email": "kudriavtseva.a@golizna.ru",
        "job": "Менеджер по закупкам",
        "default": false,
        "deleted": false,
        "phone": "+7(909)332-15-94"
      },
      {
        "id": 41,
        "client_id": 32,
        "name": "Щукин Даниил Евгеньевич",
        "email": "shchukin.d@golizna.ru",
        "job": "Менеджер по закупкам",
        "default": false,
        "deleted": false,
        "phone": "+7(974)986-82-21"
      }
    ],
    "addresses": [
      {
        "id": 44,
        "client_id": 32,
        "delivery_type": "cdek",
        "region": "Калужская",
        "city": "Подольск",
        "metro": null,
        "address": "спуск Славы, 38",
        "point_id": null,
        "cdek_type": "courier",
        "default": false,
        "deleted": false
      },
      {
        "id": 45,
        "client_id": 32,
        "delivery_type": "cdek",
        "region": "Мурманская",
        "city": "Луховицы",
        "metro": null,
        "address": "шоссе Гагарина, 21",
        "point_id": "SXT0095",
        "cdek_type": "pvz",
        "default": false,
        "deleted": false
      }
    ]
  },
  "address": {
    "id": 81,
    "client_id": 64,
    "delivery_type": "courier",
    "region": null,
    "city": null,
    "metro": "Театральная",
    "address": "спуск Гагарина, 98",
    "point_id": null,
    "cdek_type": null,
    "default": false,
    "deleted": false
  },
  "created_at": "2024-03-04T13:12:24.000Z",
  "updated_at": "2024-03-04T13:12:24.000Z",
  "manager": {
    "id": 4,
    "first_name": "Сергей",
    "last_name": "Сергеев",
    "gender": "male",
    "email": "postprinter@ottisk.com",
    "email_verified_at": null,
    "deleted": false,
    "created_at": "2024-03-04T13:12:00.000000Z",
    "updated_at": "2024-03-04T13:12:00.000000Z",
    "user_role": {
      "name": "Работник постпечати",
      "capabilities": [
        "user",
        "postprint",
        "worker",
        "services"
      ],
      "states": [
        "postprint/inbox",
        "postprint/in-progress",
        "quality-checking"
      ]
    }
  }
} as unknown as Delivery;

vi.mock("react-i18next", async () => {
  const originalModule = await vi.importActual<typeof import("react-i18next")>(
    "react-i18next"
  ); // Dynamically import the actual react-i18next module.

  return {
    ...originalModule, // Spread all exports from the original module.
    useTranslation: () => ({
      // Override the useTranslation hook as needed for testing.
      t: (key: string) => {
        if (key === "delivery.chip.expired") return "Expired";
        if (key === "delivery.chip.timeLeft") return "1h 3m";
        if (key === "delivery.label.address") return "Adress";
        if (key === "delivery.label.address.notFound") return "Undefined";
        if (key === "delivery.client.name.notFound") return "Not found";
        return key;
      },
      i18n: {
        changeLanguage: () => new Promise(() => {
        })
      }
    })
  };
});


// Mock useEstimatedTime hook
vi.mock("../../../lib/hooks/useEstimatedTime", () => ({
  useEstimatedTime: vi.fn()
}));

describe("<DeliveryCountdownCard />", () => {
  beforeEach(() => {
    // Reset mocks before each test
    vi.resetAllMocks();
  });

  it("renders correctly with valid delivery data", () => {
    // Mock useEstimatedTime to return 63 for 1h 3m
    vi.mocked(useEstimatedTime).mockReturnValue(63);

    const { asFragment } = render(<DeliveryCountdownCard delivery={mockDelivery} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("displays \"Expired\" when estimated time is negative", () => {
    // Mock useEstimatedTime to return a negative value indicating an expired time
    vi.mocked(useEstimatedTime).mockReturnValue(-10);

    render(<DeliveryCountdownCard delivery={mockDelivery} />);
    expect(screen.getByText("Expired")).toBeInTheDocument();
  });

  it("handles undefined delivery address and contact gracefully", () => {
    // Assuming 20 minutes left for delivery
    vi.mocked(useEstimatedTime).mockReturnValue(20);

    const delivery = {};

    render(<DeliveryCountdownCard delivery={delivery as unknown as Delivery} />);

    expect(screen.getByText("Undefined")).toBeInTheDocument();
    expect(screen.getByText("Not found")).toBeInTheDocument();
  });

  it("displays provided address and contact correctly", () => {
    // Custom time to avoid relying on real-time data
    vi.mocked(useEstimatedTime).mockReturnValue(45);

    render(<DeliveryCountdownCard delivery={mockDelivery} />);

    expect(screen.getByText("спуск Гагарина, 98")).toBeInTheDocument();
    expect(screen.getByText("Белозёрова Лидия Сергеевна")).toBeInTheDocument();
    expect(screen.getByText("+7(999)125-12-02")).toBeInTheDocument();
  });
});