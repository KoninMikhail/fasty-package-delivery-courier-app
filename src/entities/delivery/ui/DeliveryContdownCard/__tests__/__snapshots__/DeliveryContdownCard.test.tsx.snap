// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<DeliveryCountdownCard /> > renders correctly with valid delivery data 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col relative overflow-hidden height-auto text-foreground box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none min-w-[300px] max-w-[600px]"
    tabindex="-1"
  >
    <div
      class="p-3 z-10 w-full justify-start items-center shrink-0 overflow-inherit color-inherit subpixel-antialiased rounded-t-large flex gap-3"
    >
      <div
        class="flex w-full items-center gap-2"
      >
        <div
          class="flex-grow"
        >
          <div
            class="inline-flex items-center gap-1.5"
          >
            <img
              class="h-5 w-5"
              src="/icons/subway/Moskwa_Metro_Unknown.svg"
            />
            <span
              class="truncate"
            >
              Театральная
            </span>
          </div>
        </div>
        <div
          class="flex items-center gap-1.5"
        >
          <div
            class="relative max-w-fit inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-default text-default-foreground"
          >
            <span
              class="flex-1 text-inherit font-normal px-2"
            >
              1h 3m
            </span>
          </div>
        </div>
      </div>
    </div>
    <hr
      class="shrink-0 bg-divider border-none w-full h-divider"
      role="separator"
    />
    <div
      class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left overflow-y-auto subpixel-antialiased"
    >
      <div
        class="flex flex-col"
      >
        <span
          class="text-md"
        >
          Adress:
        </span>
        <span
          class="text-md font-bold"
        >
          спуск Гагарина, 98
        </span>
      </div>
    </div>
    <hr
      class="shrink-0 bg-divider border-none w-full h-divider"
      role="separator"
    />
    <div
      class="p-3 h-auto flex w-full items-center overflow-hidden color-inherit subpixel-antialiased rounded-b-large"
    >
      <div
        class="flex w-full justify-between gap-2"
      >
        <div
          class="flex-grow"
        >
          <div
            class="truncate text-sm font-bold"
          >
            Белозёрова Лидия Сергеевна
          </div>
          <div
            class="text-sm"
          >
            +7(999)125-12-02
          </div>
        </div>
        <a
          class="tap-highlight-transparent no-underline hover:opacity-80 active:opacity-disabled transition-opacity z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 gap-unit-2 rounded-medium px-unit-0 !gap-unit-0 data-[pressed=true]:scale-[0.97] transition-transform-colors-opacity motion-reduce:transition-none bg-success min-w-unit-10 w-unit-10 h-unit-10 data-[hover=true]:opacity-hover text-lg text-content1"
          href="tel:79991251202"
          role="link"
          tabindex="0"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M391 480c-19.52 0-46.94-7.06-88-30-49.93-28-88.55-53.85-138.21-103.38C116.91 298.77 93.61 267.79 61 208.45c-36.84-67-30.56-102.12-23.54-117.13C45.82 73.38 58.16 62.65 74.11 52a176.3 176.3 0 0 1 28.64-15.2c1-.43 1.93-.84 2.76-1.21 4.95-2.23 12.45-5.6 21.95-2 6.34 2.38 12 7.25 20.86 16 18.17 17.92 43 57.83 52.16 77.43 6.15 13.21 10.22 21.93 10.23 31.71 0 11.45-5.76 20.28-12.75 29.81-1.31 1.79-2.61 3.5-3.87 5.16-7.61 10-9.28 12.89-8.18 18.05 2.23 10.37 18.86 41.24 46.19 68.51s57.31 42.85 67.72 45.07c5.38 1.15 8.33-.59 18.65-8.47 1.48-1.13 3-2.3 4.59-3.47 10.66-7.93 19.08-13.54 30.26-13.54h.06c9.73 0 18.06 4.22 31.86 11.18 18 9.08 59.11 33.59 77.14 51.78 8.77 8.84 13.66 14.48 16.05 20.81 3.6 9.53.21 17-2 22-.37.83-.78 1.74-1.21 2.75a176.49 176.49 0 0 1-15.29 28.58c-10.63 15.9-21.4 28.21-39.38 36.58A67.42 67.42 0 0 1 391 480z"
            />
          </svg>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
